newrelic_lambda_cli/integrations.py

```diff
@@ -76,93 +79,256 @@ def create_role(session, role_policy, nr_account_id):
+
+    change_set = cf_client.create_change_set(
+        StackName=INGEST_STACK_NAME,
+        TemplateURL=template_url,
+        Parameters=parameters,
+        Capabilities=capabilities,
+        ChangeSetType=mode,
+        ChangeSetName=change_set_name,
+    )
+
+    exec_change_set(cf_client, change_set, mode)
+
+
+def exec_change_set(cf_client, change_set, mode):
+    click.echo(
+        "Waiting for change set creation to complete, this may take a minute... "
+    )
+    cf_client.get_waiter("change_set_create_complete").wait(
+        ChangeSetName=change_set["Id"], WaiterConfig={"Delay": 10},
+    )
+    cf_client.execute_change_set(ChangeSetName=change_set["Id"])
+    click.echo(
+        "Waiting for change set to finish execution. This may take a minute... ",
+        nl=False,
+    )
+    exec_waiter_type = f"stack_{mode.lower()}_complete"
+    cf_client.get_waiter(exec_waiter_type).wait(
+        StackName=INGEST_STACK_NAME, WaiterConfig={"Delay": 15},
+    )
+    success("Done")
 
 
 def update_log_ingestion_function(
-    session, nr_license_key, enable_logs=False, memory_size=128, timeout=30
+    session, nr_license_key, enable_logs, memory_size, timeout, role_name=None,
 ):
+    # Detect an old-style nested install and unwrap it
     client = session.client("cloudformation")
-    stack_name = "NewRelicLogIngestion"
-    template_path = os.path.join(
-        os.path.dirname(os.path.abspath(__file__)),
-        "templates",
-        "newrelic-log-ingestion.yaml",
+
+    resources = client.describe_stack_resources(
+        StackName=INGEST_STACK_NAME, LogicalResourceId="NewRelicLogIngestion"
     )
-    with open(template_path) as template:
+    stack_resources = resources["StackResources"]
+    # The nested installs had a single Application resource
+    if (
+        len(stack_resources) > 0
+        and stack_resources[0]["ResourceType"] == "AWS::CloudFormation::Stack"
+    ):
+        click.echo("Unwrapping nested stack")
+
+        # Set the ingest function itself to disallow deletes
+        nested_stack = stack_resources[0]
+        template_response = client.get_template(
+            StackName=nested_stack["PhysicalResourceId"], TemplateStage="Processed"
+        )
+        template_body = template_response["TemplateBody"]
+        template_body["Resources"]["NewRelicLogIngestionFunction"][
+            "DeletionPolicy"
+        ] = "Retain"
+        template_body["Resources"]["NewRelicLogIngestionFunctionRole"][
+            "DeletionPolicy"
+        ] = "Retain"
+
+        # We can't change props during import, so let's set them to their current values
+        lambda_client = session.client("lambda")
+        old_props = lambda_client.get_function_configuration(
+            FunctionName="newrelic-log-ingestion"
+        )
+        old_role_name = old_props["Role"].split(":role/")[-1]
+        old_nr_license_key = old_props["Environment"]["Variables"]["LICENSE_KEY"]
```

Comment:
Should we preserve any other environment variables the user has set on the log ingestion function? The new template here only has a subset of the environment variables that can be set. I'm wondering if we should expose all the supported environment variables so they don't get clobbered during an update. Thinking specifically about debug mode and endpoint overrides.

Reply:
This mechanism only works with props. We're using the env vars as a source for prop values, because that works with our existing templates.

In general, there isn't a way to preserve arbitrary env vars across template updates, as far as I can tell.

Reply:
Testing the upgrade with an older function that has additional env vars, it looks like those vars get preserved. So I think we're OK here.